calico/felix/ipsets.py

```diff
@@ -544,12 +544,16 @@ def _finish_msg_batch(self, batch, results):
             self._force_reprogram = False
 
     def _sync_to_ipset(self):
-        _log.info("Rewriting %s with %d members.", self, len(self.members))
         _log.debug("Setting ipset %s to %s", self, self.members)
-        # Defer to our helper.
-        self._ipset.replace_members(self.members)
-        # We have got the set into the correct state.
-        self.programmed_members = self.members.copy()
+        # Defer to our helper to actually make the changes.
+        if self._force_reprogram or self.programmed_members is None:
+            self._ipset.replace_members(self.members)
+            self.programmed_members = self.members.copy()
```

Comment:
> I wish Python had an immutable set type

What about frozenset? Looks like it implements a very similar API to set(), except it's immutable and hashable. No idea how using frozenset affects occupancy.

(Standard library, apparently available at least as far back as 2.6)